File: shared/database/models/post.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type PostType string

const (
	Lost  PostType = "lost"
	Found PostType = "found"
)

type Post struct {
	ID                string     `gorm:"primaryKey" json:"id"`
	CreatedAt         time.Time  `json:"created_at"`
	TimeAgo           *string    `gorm:"-:migration;->" json:"time_ago,omitempty"`
	UpdatedAt         time.Time  `json:"updated_at"`
	Title             string     `json:"title"`
	Description       *string    `json:"description"`
	Time              *time.Time `gorm:"default:CURRENT_TIMESTAMP" json:"time,omitempty"`
	LocationLatitude  *float64   `json:"location_latitude"`
	LocationLongitude *float64   `json:"location_longitude"`
	Type              PostType   `sql:"type:ENUM('lost', 'found')" gorm:"default:'lost'" json:"type"`
	HasBeenFound      bool       `gorm:"default:false" json:"has_been_found"`
	HasBeenDelivered  bool       `gorm:"default:false" json:"has_been_delivered"`
	ClaimedByUser     *bool      `gorm:"-:migration;->" json:"claimed_by_user,omitempty"`
	FoundByUser       *bool      `gorm:"-:migration;->" json:"found_by_user,omitempty"`
	UserID            string     `json:"user_id"`
	User              *User      `json:"user,omitempty"`
	Images            []Image    `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"images"`
	ClaimersCount     *uint      `gorm:"-:migration;->" json:"claimers_count,omitempty"`
	Claimers          []User     `gorm:"many2many:claims" json:"claimers,omitempty"`
	FoundersCount     *uint      `gorm:"-:migration;->" json:"founders_count,omitempty"`
	Founders          []User     `gorm:"many2many:founds" json:"founders,omitempty"`
}

func (p *Post) BeforeCreate(tx *gorm.DB) (err error) {
	p.ID = uuid.New().String()
	return
}

type Image struct {
	ID                string `gorm:"primaryKey" json:"id"`
	Name              string `gorm:"not null" json:"name"`
	URL               string `gorm:"unique,not null" json:"url"`
	FileStorageFolder string `json:"file_storage_folder"`
	PostID            string `json:"post_id"`
	Post              *Post  `json:"post,omitempty"`
}

func (i *Image) BeforeCreate(tx *gorm.DB) (err error) {
	i.ID = uuid.New().String()
	return
}

File: lib/shared/models/post.dart
import 'package:estin_losts/shared/utils/utils.dart';
import 'package:estin_losts/shared/models/user.dart';

enum PostType { lost, found }

class Post {
  final String id;
  final DateTime createdAt;
  final String? timeAgo;
  final DateTime updatedAt;
  final String title;
  final String? description;
  final double? locationLatitude;
  final double? locationLongitude;
  final String? locationDescription;
  final PostType type;
  final bool hasBeenFound;
  final bool hasBeenDelivered;
  final bool? claimedByUser;
  final bool? foundByUser;
  final String userID;
  final User? user;
  final List<PostImage> images;
  final int claimersCount;
  final List<User> claimers;
  final int foundersCount;
  final List<User> founders;

  Post({
    required this.id,
    required this.createdAt,
    this.timeAgo,
    required this.updatedAt,
    required this.title,
    this.description,
    this.locationLatitude,
    this.locationLongitude,
    this.locationDescription,
    required this.type,
    required this.hasBeenFound,
    required this.hasBeenDelivered,
    this.claimedByUser,
    this.foundByUser,
    required this.userID,
    this.user,
    this.images = const [],
    this.claimersCount = 0,
    this.claimers = const [],
    this.foundersCount = 0,
    this.founders = const [],
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: json['id'],
      createdAt: DateTime.parse(json['created_at']),
      timeAgo: json['time_ago'],
      updatedAt: DateTime.parse(json['updated_at']),
      title: json['title'],
      description: json['description'] ?? "",
      locationLatitude: json['location_latitude'] ?? 0.0,
      locationLongitude: json['location_longitude'] ?? 0.0,
      locationDescription: json['location_description'],
      type: json['type'] as String == 'lost' 
        ? PostType.lost 
        : PostType.found,  
      hasBeenFound: json['has_been_found'],
      hasBeenDelivered: json['has_been_delivered'],
      claimedByUser: json['claimed_by_user'],
      foundByUser: json['found_by_user'],
      userID: json['user_id'],
      user: json['user'] != null ? User.fromMap(json['user']) : null,
      images: ((json['images'] ?? []) as List<dynamic>)
        .map((imageJson) => PostImage.fromJson(imageJson))
        .toList(),
      claimersCount: json['claimers_count'] ?? 0,
      claimers: ((json['claimers'] ?? []) as List<dynamic>)
        .map((imageJson) => User.fromMap(imageJson))
        .toList(),
      foundersCount: json['founders_count'] ?? 0,
      founders: ((json['founders'] ?? []) as List<dynamic>)
        .map((imageJson) => User.fromMap(imageJson))
        .toList(),
    );
  }
}

class PostImage {
  final String id;
  final String name;
  String url;
  final String fileStorageFolder;
  final String postID;
  final Post? post;

  PostImage({
    required this.url,
    required this.id,
    required this.name,
    required this.fileStorageFolder,
    required this.postID,
    this.post,
  }){
    if (url  == "") {
      url = Utils.getRandomImageURL();
    } else {
      url = "http://$url";
    }
  }

  factory PostImage.fromJson(Map<String, dynamic> json) {
    return PostImage(
      id: json['id'],
      name: json['name'],
      url: json['url'],
      fileStorageFolder: json['file_storage_folder'],
      postID: json['post_id'],
      post: json['post'],
    );
  }
}